src/main/java/org/spongepowered/common/mixin/core/network/MixinNetHandlerPlayServer.java

```diff
@@ -421,6 +425,73 @@ public void processCreativeInventoryAction(CPacketCreativeInventoryAction packet
         }
     }
 
+    /**
+     * @reason When a player use the spectator menu to teleport anyone,
+     * Sponge should fire the MoveEntityEvent.Teleport.
+     *
+     * @param packetIn The spectator packet
+     */
+    @Overwrite
+    public void handleSpectate(CPacketSpectate packetIn) {
+        PacketThreadUtil.checkThreadAndEnqueue(packetIn, (NetHandlerPlayServer) (Object) this, this.player.getServerWorld());
+
+        if (this.player.isSpectator()) {
+            Entity entity = null;
+
+            for (WorldServer worldserver : this.server.worlds) {
+                if (worldserver != null) {
+                    entity = packetIn.getEntity(worldserver);
+
+                    if (entity != null) {
+                        break;
+                    }
+                }
+            }
+
+            if (entity != null) {
+                this.player.setSpectatingEntity(this.player);
+                this.player.dismountRidingEntity();
+                // Sponge start - handle MoveEntityEvent.Teleport
+                double x = entity.posX;
```

Comment:
If all that is being done is based on the player, and entity, you should be able to reduce this to a single cancellable injection:
```java
    @Inject(
        method = "handleSpectate",
        at = @At(
            value = "FIELD",
            target = "Lnet/minecraft/entity/Entity;world:Lnet/minecraft/world/World;",
            ordinal = 0
        ),
        locals = LocalCapture.CAPTURE_FAILHARD,
        cancellable = true
    )
    private void onSpectateTeleportForSponge(CPacketSpectate packetIn, CallbackInfo ci, Entity spectatingEntity) {
        double x = spectatingEntity.posX;
        double y = spectatingEntity.posY;
        double z = spectatingEntity.posZ;
        float yaw = spectatingEntity.rotationYaw;
        float pitch = spectatingEntity.rotationPitch;
        MoveEntityEvent.Teleport event = EntityUtil.handleDisplaceEntityTeleportEvent(this.player, x, y, z, yaw, pitch);
        if (event.isCancelled()) {
            ci.cancel();
        }

    }
```
Also, make sure your javadoc includes `@author` and the date, and the version of Minecraft targeted.